lib/posix-vfs-fstab: Introduce fstab library

andreittr opened the pull request:
### Description of changes

This change introduces posix-vfs-fstab, a library tasked with mounting filesystems at boot according to defined tables (fstabs).
Its general functionality mirrors that of legacy vfscore automount.

As a prerequisite the build system now supports `Config.uk.d` drop-in directories for Kconfig files in addition to the required `Config.uk`. This allows libraries (like this one) to better parcel out complex or verbose config options.

~~Depends on the following PR and has it included as a squashed commit; ignore during review:~~
~~- https://github.com/unikraft/unikraft/pull/1610~~

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

Config options for setting the builtin fstab(s) are similar to but not the same as the ones used by vfscore's `automount`. Users need to reconfigure their kernels, and adapt any scripts that emit `.config` files accordingly.

Legacy vfscore's automount functionality remains unchanged and orthogonally available.

andreittr committed:
lib/posix-vfs-fstab: Introduce fstab library

This change introduces posix-vfs-fstab, a library tasked with mounting
filesystems at boot according to defined tables (fstabs).
Its general functionality mirrors that of legacy vfscore automount.

Checkpatch-Ignore: ENOSYS
Checkpatch-Ignore: LONG_LINE_COMMENT
Signed-off-by: Andrei Tatar <andrei@unikraft.io>

mogasergiu commented on support/build/config-submenu.sh:
```diff
@@ -160,6 +160,15 @@ esac
+
+	CONFIG_UK_D="${CONFIG_UK}.d"
+	if [ -d "${CONFIG_UK_D}" ]; then
+		IFS=$IFS_NL
```
Any reason you did not use `{ }` for consistency for the `IFS_NL` and `IFS_ORIG` vars?

andreittr replied:
Force of habit most likely; but on 2nd look I don't like this invasive change to the buildsystem for something that is likely only to be used by libraries that do crazy `Config.uk` wizardry, which I'm hoping there won't be many of.
Will drop this commit in the pass and replace with bespoke `Config.uk` magic in `posix-vfs-fstab` itself.

edit: geez, I realised I had missed to include the `Config.uk.d` directory for fstab entirely :sweat_smile: .

mogasergiu commented on lib/posix-vfs-fstab/fstab-builtin.h:
```diff
@@ -0,0 +1,184 @@
+const struct vfs_fstab_entry fstab_bi0 = {
+#ifdef CONFIG_LIBPOSIX_VFS_FSTAB_BI0_DEV_ARG
+	.dev = CONFIG_LIBPOSIX_VFS_FSTAB_BI0_DEV_ARG,
+#else
```
`!CONFIG_LIBPOSIX_VFS_FSTAB_BI0_DEV_ARG` and ditto for `endif`? (Ditto for the others in this strucct as well)

mogasergiu commented on lib/posix-vfs-fstab/einitrd.S:
```diff
@@ -0,0 +1,19 @@
+
+#include <uk/config.h>
+
+.globl vfs_einitrd_start
```
Since there is no `exportsyms.uk` and I assume you don't want these syms exported, consider adding a `localsyms.uk` file.

mogasergiu commented on lib/posix-vfs-fstab/fstab.c:
```diff
@@ -0,0 +1,353 @@
+	int mounted = 0;
+
+	for (size_t i = 0; i < len; i++)
+		if (likely(fstab[i])) {
```
You'll hate me for mentioning it :sweat_smile: but I think an `unlikely` with `continue` and reduced indentation here would look better? But is possible for an entry to be NULL? Did you mean to compare against `""` instead?


(Ditto `dddfstab_process_user`)

andreittr replied:
> But is possible for an entry to be NULL?

[Yes](https://github.com/unikraft/unikraft/pull/1654/files/5f887d1c7641ba7eaeb765b700a2a950ac7de9e9..19e98078083e6a4b54bee75849505f082cf93ecc) although I'm not sure we need a `NULL` sentinel value if we know the array size :thinking: could probably rework this.

edit: and on closer look, user fstabs can have multiple NULL entries at the end, since it's a fixed-size array that may get populated by more or fewer entries at runtime. But there we should probably break the loop at the first NULL :thinking: .

Regardless, I'll remove branch hinting, as it doesn't denote an error condition or anything that would warrant a "cold path".

mogasergiu commented on lib/posix-vfs-fstab/fstab.c:
```diff
@@ -0,0 +1,353 @@
+			ret |= FSTAB_UKOPT_MKMP;
+			continue;
+		}
+		if (!strcmp(opt, "ifinitrd0")) {
```
Maybe we could use `strncmp` with `sizeof("hardcoded string here")` instead? (Ditto for `fstab_mount`/`fstab_extract`'s already macronized strings)

andreittr replied:
I disagree; using `strncmp` with string literals would only harm things here: it requires either duplicating the literal for `sizeof` (which introduces the risk of divergence and hard-to-catch bugs) or macro noise to define the length of every literal we care to compare against.

`strcmp` is IMO fine here because bounds checking is guaranteed by the finite length of the string literal; `strncmp` would add nothing in terms of safety. And if you're worried about performance, any halfway decent compiler will LTO a `strcmp` where one of the strings is a literal to the unrolled equivalent
```c
if (str[0] && str[0] == lit[0] && str[1] && str[1] == lit[1] && ...)
```

mogasergiu commented on lib/posix-vfs-fstab/fstab.c:
```diff
@@ -0,0 +1,353 @@
+				break; /* End of option string */
+			continue; /* Empty option */
+		}
+		if (!strcmp(opt, "mkmp")) {
```
I think these commands could be macronized instead of being hardcoded here?

mogasergiu commented on lib/posix-vfs-fstab/fstab.c:
```diff
@@ -0,0 +1,353 @@
+		uk_pr_crit("Invalid initrd source: %s\n", e->dev);
+		return -EINVAL;
+	}
+
```
Would it make sense to have two `UK_ASSERT`s here: one for `base` and one for `len`?

andreittr replied:
Not at all, I think it would make things more brittle. The code here in fstab is perfectly correct and consistent regardless of their values. How `ukcpio_extract` uses those values passed as args is none of our business here in fstab ([its API docs](https://github.com/unikraft/unikraft/blob/staging/lib/ukcpio/include/uk/cpio.h#L163) say nothing about zero values being forbidden, and [the implementation](https://github.com/unikraft/unikraft/blob/staging/lib/ukcpio/cpio.c#L333) doesn't seem to mind either).

I don't see what benefit we gain from changing previously OK behavior (extracting a 0-length CPIO is a NOP) to "illegal state we must crash the kernel".

mogasergiu commented on lib/posix-vfs-fstab/fstab.c:
```diff
@@ -0,0 +1,353 @@
+ * INTERNAL. Get the memregion of initrd0.
+ *
+ * @return
+ *  Zero on success, non-zero on failure.
```
This is documented as if it would return an integer :sweat_smile:

mogasergiu commented on lib/posix-vfs-fstab/fstab.c:
```diff
@@ -0,0 +1,353 @@
+		e->flags = strtol(flags_str, NULL, 0);
+	else
+		e->flags = 0;
+	e->ukopts = fstab_parse_ukopts(ukopts_str);
```
I think we should have a check/assert somewhere for `ukopts_str`? Either here or in `fstab_parse_ukopts` :thinking:

andreittr replied:
What for? A null value is handled perfectly gracefully by `fstab_parse_ukopts` and `uk_nextarg`. No need to crash the kernel out of paranoia.

That said, I see a logic error in how `ukopts_str` is assigned, since both vol and ukopts share the same separator char, the line
```c
ukopts_str = uk_nextarg(&p, FSTAB_VOL_ARGS_SEP);
```
would chop off all but the first ukopt before we got into `fstab_parse_ukopts`. Thus it should simply be
```c
ukopts_str = p;
```
and let `fstab_parse_ukopts` parse the rest until `\0`.

mogasergiu commented on lib/posix-vfs-fstab/fstab.c:
```diff
@@ -0,0 +1,353 @@
+		e->fstype = "";
+
+	if (flags_str && flags_str[0])
+		e->flags = strtol(flags_str, NULL, 0);
```
`strtol`.. which also sets errno, should we check for underflow/overflow?

andreittr replied:
There's not much that fstab can meaningfully do on underflow/overflow except crash, and I'm not sure whether that's the right thing to do here :shrug: . But ignoring the error is probably not kosher either. I'll add some error checking and warnings.

Re: `strtol`, yes, it's the only function that does multi-base conversion without having to roll my own. And its touching errno is irrelevant, as everybody is allowed to set errno, [even functions that succeed](https://www.man7.org/linux/man-pages/man3/errno.3.html#DESCRIPTION).

mogasergiu commented on lib/posix-vfs-fstab/fstab.c:
```diff
@@ -0,0 +1,353 @@
+		if (unlikely(r))
+			uk_pr_crit("Coult not create mount path '%s': %d\n",
+				   e->path, r);
+		/* Attempt to mount even on failure */
```
Can you also mention the reason for this?

mogasergiu commented on lib/posix-vfs-fstab/fstab.c:
```diff
@@ -0,0 +1,353 @@
+	if (e->ukopts & FSTAB_UKOPT_MKMP) {
+		r = fstab_mkmp(e->path);
+		if (unlikely(r))
+			uk_pr_crit("Coult not create mount path '%s': %d\n",
```
s/Coult/Could

mogasergiu reviewed (changes requested):
A couple of comments/questions.

andreittr commented:
Pushed rebase to staging & latest #1610 .

andreittr commented:
Pushed addressing comments + more:
- dropped changes to build system; usecase not common enough to warrant it
- added missing builtin & user fstab config dropins (oops)
- added missing library registration in `lib/Makefile.uk` (oops)
- builtin and fallback fstabs have had one layer of indirection removed; they are now flat struct arrays in the kernel image
- misc code improvements

andreittr commented:
Pushed cosmetic changes to fix checkpatch warnings.

mogasergiu commented on lib/posix-vfs-fstab/user/Config.uk:
```diff
@@ -0,0 +1,460 @@
+
+	config LIBPOSIX_VFS_FSTAB_FALLBACK_EINITRD
+	bool "RAMfs + compiled-in CPIO initrd"
+	select LIBPOSIX_VFS_FSTAB_EINITRD
```
I am not sure how consistent the KConfig style is. Normally you don't indent these (yes i know we have that in some places but there's been changes in the right direction). Additionally, the help text in some places seems to be inconsistent and does not always begin 2 whitespaces indentation on the next line after `help` (e.g. `LIBPOSIX_VFS_FSTAB_FALLBACK_EINITRD_EXTRACT`). The style that I think is commonly used is
```
config NAME
<tab>bool/prompt/etc
<tab>select/depends on/etc
<tab>help
<tab>  <message here>
```
(You can look at `lib/ukboot/Config.uk` for example)

andreittr replied:
Good point, I had blindly copied over an older style, will update :+1:

andreittr commented:
Pushed rebase to current staging & latest #1610 .

andreittr commented:
Pushed update addressing comment & clarifying fstab's intentional positioning as latest in the rootfs init class.

mogasergiu commented on lib/posix-vfs-fstab/user/Config.uk:
```diff
@@ -0,0 +1,460 @@
+	select LIBPOSIX_VFS_FSTAB_EINITRD
+	select LIBUKCPIO
+	help
+		Extract a compiled-in initrd into an existing directory.
```
Commenting so we don't forget: indent this line properly.

mogasergiu reviewed (approved):
LGTM just 1 more comment.

Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

andreittr commented:
Pushed update adding `exportsyms.uk` file to explicitly express developer intentions, and remove a `localsyms.uk` file that was mistakenly put in the repo root :sweat_smile: .

andreittr commented:
Pushed update addressing final comment.

andreittr commented:
Pushed rebase to current staging following merging of all prerequisites.

michpappas commented:
Approved-by: Michalis Pappas <michalis@unikraft.io>
